docs/instrumentation/tracing-java-interop.md

```diff
@@ -190,8 +186,30 @@ Java ctx: {traceId=06f5d9112efbe711947ebbded1287a30, spanId=26ed80c398cc039f, ..
 Otel4s ctx: {traceId=06f5d9112efbe711947ebbded1287a30, spanId=26ed80c398cc039f, ...}
 ```
 
-As we can see, the tracing information is in sync now,
-and you can use Java-instrumented libraries within the `useJContext` block. 
+As we can see, the tracing information is now available in ThreadLocal now too.
+Code instrumented using OpenTelemetry's Java API will work inside `blockingWithContext`.
+
+### Calling asynchronous code
+
+When interopting with asynchronous Java APIs:
+
+```
+  def asyncWithContext[F[_]: Async, A](k: (Either[Throwable, A] => Unit) => F[Option[F[Unit]]])(implicit L: Local[F, Context]): F[A] =
+    Local[F, Context].ask.flatMap { ctx =>      // <1>
+      Async[F].async[A] { cb =>
+        Async[F].delay {
+          val jContext: JContext = ctx.underlying
+          val _ = jContext.makeCurrent()
+        }.flatMap(_ => k(cb))
```

Comment:
I think this is problematic because there's no guarantee that delay and the subsequent flatMap are executed together. Any ideas?

Reply:
Perhaps we can use `evalOn` with a custom `ExecutonContext`:

```scala
def tracedContext(ctx: JContext): ExecutionContext =
  new ExecutionContext {
    def execute(runnable: Runnable): Unit = {
      val scope = ctx.makeCurrent()
      try {
        runnable.run()
      } finally {
        scope.close()
      }
    }
    def reportFailure(cause: Throwable): Unit =
      cause.printStackTrace()
  }
  
def asyncWithContext[F[_]: Async, A](k: (Either[Throwable, A] => Unit) => F[Option[F[Unit]]])(implicit L: Local[F, Context]): F[A] =
  Local[F, Context].ask.flatMap { ctx =>
    Async[F].evalOn(
      Async[F].async[A](cb => k(cb)), 
      tracedContext(ctx.underlying)
    )
  }
```

Reply:
But implications on performance are unknown :(

Reply:
`evalOn` is applicable for all scenarios, actually. But again, it may (and highly likely will) affect the performance. 

```scala
def withJContext[F[_]: Async, A](fa: F[A])(implicit L: Local[F, Context]): F[A] = 
  Local[F, Context].ask.flatMap { ctx =>
    Async[F].evalOn(fa, tracedContext(ctx.underlying))
  }
```

Reply:
Good idea :) Why do you think it'll affect performance (assuming `fa` pretty much immediately calls the async API instead of doing a bunch of CPU work)

Reply:
AFAIK context shifts may be costly. 

The current implementation of the `tracedContext` is basically a same-thread execution.

Here is the logic for the `evalOn` in the CE: https://github.com/typelevel/cats-effect/blob/3cf97ae4fd643f333febb4554af2fb603ed2c9d2/core/shared/src/main/scala/cats/effect/IOFiber.scala#L976-L988.

If I get it right, `fa` will be executed right within the fiber's loop. 

@armanbilge perhaps you have some insides?

Reply:
I've added your suggestion for now. I'm thinking that a typical async API involves using another threadpool so context shift isn't avoidable in that case. (But yeah will be good if an additional context shift can be avoided)